spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/ComputeTableStatsSparkAction.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.iceberg.actions.ComputeTableStats;
+import org.apache.iceberg.actions.ImmutableComputeTableStats;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.puffin.Blob;
+import org.apache.iceberg.puffin.Puffin;
+import org.apache.iceberg.puffin.PuffinWriter;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.spark.JobGroupInfo;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.SparkSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Computes the statistic of the given columns and stores it as Puffin files. */
+public class ComputeTableStatsSparkAction extends BaseSparkAction<ComputeTableStatsSparkAction>
+    implements ComputeTableStats {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeTableStatsSparkAction.class);
+
+  private final Table table;
+  private Set<String> columns;
+  private long snapshotId;
+
+  ComputeTableStatsSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+    Snapshot snapshot = table.currentSnapshot();
+    if (snapshot != null) {
+      this.snapshotId = snapshot.snapshotId();
+    }
+    this.columns =
+        table.schema().columns().stream().map(Types.NestedField::name).collect(Collectors.toSet());
+  }
+
+  @Override
+  protected ComputeTableStatsSparkAction self() {
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    String desc =
+        String.format("Computing stats for %s for snapshot id %s", table.name(), snapshotId);
+    JobGroupInfo info = newJobGroupInfo("COMPUTE-TABLE-STATS", desc);
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    if (snapshotId == 0L) {
+      return ImmutableComputeTableStats.Result.builder().build();
+    }
+    LOG.info("Computing stats of {} for snapshot {}", table.name(), snapshotId);
+    List<Blob> blobs = generateNDVBlobs();
+    StatisticsFile statisticFile;
+    try {
+      statisticFile = writeAndCommitPuffin(blobs);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    return ImmutableComputeTableStats.Result.builder().statisticsFile(statisticFile).build();
+  }
+
+  private StatisticsFile writeAndCommitPuffin(List<Blob> blobs) throws Exception {
+    LOG.info("Writing stats to puffin files for table {}", table.name());
+    TableOperations operations = ((HasTableOperations) table).operations();
+    FileIO fileIO = operations.io();
+    String path = operations.metadataFileLocation(String.format("%s.stats", UUID.randomUUID()));
+    OutputFile outputFile = fileIO.newOutputFile(path);
+    GenericStatisticsFile statisticsFile;
```

Comment:
Nit: how about an extra method, to avoid having the weird declaration/assignment inside the try clause:

```
GenericStatisticsFile statisticsFile = writePuffinFile();
```

```
writePuffinFile() {
    try (PuffinWriter writer =
        Puffin.write(outputFile).createdBy("Iceberg ComputeTableStats action").build()) {
      blobs.forEach(writer::add);
      writer.finish();
      return new GenericStatisticsFile(
              snapshotId,
              path,
              writer.fileSize(),
              writer.footerSize(),
              writer.writtenBlobsMetadata().stream()
                  .map(GenericBlobMetadata::from)
                  .collect(ImmutableList.toImmutableList()));
    }
```